Provider crash on upgrade to `1.7.3`

**Describe the bug**
Provider panic's and crashes during `terraform plan` on version 1.7.3. Downgrading to 1.7.2 works.

Based on the log it seems to be related to this version trying to null out the groups value on any `platform_permission` that do not define groups. Here is an example from ours

```terraform
resource "platform_permission" "npm_remote_repo" {
  name = "npm-remote-repo"

  artifact = {
    actions = {
      users = [
        {
          name        = artifactory_user.service_account["my-account"].name
          permissions = ["READ", "WRITE"]
        }
      ]
    }

    targets = [{ name = "my-remote" }]
  }
}
```

**Expected behavior**
The provider should not crash.

**Additional context**
```
Stack trace from the terraform-provider-platform_v1.7.3 plugin:
panic: interface conversion: attr.Value is nil, not basetypes.ObjectValue
goroutine 32 [running]:
github.com/jfrog/terraform-provider-platform/pkg/platform.setUsersGroupsAttributeToNull({0xd87f08, 0xc000686930}, {0x0?, 0xc0007379e0?, 0x20?}, {0xc45b15, 0x5}, 0xc000551680)
	[github.com/jfrog/terraform-provider-platform/pkg/platform/resource_permission.go:278](http://github.com/jfrog/terraform-provider-platform/pkg/platform/resource_permission.go:278) +0x946
[github.com/jfrog/terraform-provider-platform/pkg/platform.(*permissionResource](http://github.com/jfrog/terraform-provider-platform/pkg/platform.(*permissionResource)).UpgradeState.func1({0xd87f08, 0xc000686930}, {0xc00060e0c0?, 0xc0007374d0?}, 0xc000551680)
	[github.com/jfrog/terraform-provider-platform/pkg/platform/resource_permission.go:323](http://github.com/jfrog/terraform-provider-platform/pkg/platform/resource_permission.go:323) +0x335
[github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server](http://github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server)).UpgradeResourceState(0xc00022d040, {0xd87f08, 0xc000686930}, 0xc0006869c0, 0xc00046d6d8)
	[github.com/hashicorp/terraform-plugin-framework@v1.8.0/internal/fwserver/server_upgraderesourcestate.go:205](http://github.com/hashicorp/terraform-plugin-framework@v1.8.0/internal/fwserver/server_upgraderesourcestate.go:205) +0x96c
[github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server](http://github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server)).UpgradeResourceState(0xc00022d040, {0xd87f08?, 0xc000686810?}, 0xc00060e0e0)
	[github.com/hashicorp/terraform-plugin-framework@v1.8.0/internal/proto6server/server_upgraderesourcestate.go:51](http://github.com/hashicorp/terraform-plugin-framework@v1.8.0/internal/proto6server/server_upgraderesourcestate.go:51) +0x330
[github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server](http://github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server)).UpgradeResourceState(0xc000241cc0, {0xd87f08?, 0xc000686090?}, 0xc00068e000)
	[github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/tf6server/server.go:744](http://github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/tf6server/server.go:744) +0x278
[github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_UpgradeResourceState_Handler](http://github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_UpgradeResourceState_Handler)({0xc220e0?, 0xc000241cc0}, {0xd87f08, 0xc000686090}, 0xc00068c000, 0x0)
	[github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:446](http://github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:446) +0x169
[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).processUnaryRPC(0xc0001af000, {0xd87f08, 0xc000686000}, {0xd8cf18, 0xc0001c3e00}, 0xc0003c25a0, 0xc0003401b0, 0x127c8b8, 0x0)
	[google.golang.org/grpc@v1.63.2/server.go:1369](http://google.golang.org/grpc@v1.63.2/server.go:1369) +0xe23
[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).handleStream(0xc0001af000, {0xd8cf18, 0xc0001c3e00}, 0xc0003c25a0)
	[google.golang.org/grpc@v1.63.2/server.go:1780](http://google.golang.org/grpc@v1.63.2/server.go:1780) +0x1016
[google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).serveStreams.func2.1()
	[google.golang.org/grpc@v1.63.2/server.go:1019](http://google.golang.org/grpc@v1.63.2/server.go:1019) +0x8b
created by [google.golang.org/grpc.(*Server](http://google.golang.org/grpc.(*Server)).serveStreams.func2 in goroutine 13
	[google.golang.org/grpc@v1.63.2/server.go:1030](http://google.golang.org/grpc@v1.63.2/server.go:1030) +0x135
Error: The terraform-provider-platform_v1.7.3 plugin crashed!
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,9 @@
+## 1.7.4 (May 8, 2024)
+
+BUG FIXES:
+
+* resource/platform_permission: Fix state upgrader crash when upgrading from 1.7.2 to 1.7.3 with no `groups` or `users` attribute set. Issue: [#75](https://github.com/jfrog/terraform-provider-platform/issues/75) PR: [#76](https://github.com/jfrog/terraform-provider-platform/pull/76)
+
 ## 1.7.3 (May 7, 2024)
 
 BUG FIXES:

--- a/pkg/platform/resource_permission.go
+++ b/pkg/platform/resource_permission.go
@@ -21,6 +21,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/jfrog/terraform-provider-shared/util"
 	utilfw "github.com/jfrog/terraform-provider-shared/util/fw"
 	"github.com/samber/lo"
@@ -273,23 +274,68 @@ func (r *permissionResource) Schema(ctx context.Context, req resource.SchemaRequ
 	}
 }
 
-func setUsersGroupsAttributeToNull(ctx context.Context, resource types.Object, resourceName string, resp *resource.UpgradeStateResponse) {
+func setUsersGroupsAttributeToNull(ctx context.Context, resource types.Object, resourceName string, resp *resource.UpgradeStateResponse) diag.Diagnostics {
+	if resource.IsNull() {
+		return nil
+	}
+
 	attrs := resource.Attributes()
-	actionsAttrs := attrs["actions"].(types.Object).Attributes()
-	usersSet := actionsAttrs["users"].(types.Set)
+	a, ok := attrs["actions"]
+	if !ok {
+		return nil
+	}
+	actionsAttrs := a.(types.Object).Attributes()
 
 	// actions.users and actions.groups no longer allows to be empty set.
 	// they can either be null (not set) or set with items
 	// When prior state has empty set then migrates the value to null
+	ds := diag.Diagnostics{}
 
-	if !usersSet.IsNull() && len(usersSet.Elements()) == 0 {
-		resp.State.SetAttribute(ctx, path.Root(fmt.Sprintf("%s.actions.users", resourceName)), types.SetNull(usersGroupsResourceModelAttributeTypes))
+	u, ok := actionsAttrs["users"]
+	if !ok {
+		return nil
+	}
+	ds.Append(setAttributeToNull(ctx, u.(types.Set), resourceName, "users", resp)...)
+	if ds.HasError() {
+		return ds
 	}
 
-	groupsSet := actionsAttrs["groups"].(types.Set)
-	if !groupsSet.IsNull() && len(groupsSet.Elements()) == 0 {
-		resp.State.SetAttribute(ctx, path.Root(fmt.Sprintf("%s.actions.groups", resourceName)), types.SetNull(usersGroupsResourceModelAttributeTypes))
+	g, ok := actionsAttrs["groups"]
+	if !ok {
+		return nil
 	}
+	ds.Append(setAttributeToNull(ctx, g.(types.Set), resourceName, "groups", resp)...)
+	if ds.HasError() {
+		return ds
+	}
+
+	return ds
+}
+
+func setAttributeToNull(ctx context.Context, stateSet types.Set, resourceName, attrName string, resp *resource.UpgradeStateResponse) diag.Diagnostics {
+	ds := diag.Diagnostics{}
+	if !stateSet.IsNull() && len(stateSet.Elements()) == 0 {
+		// find the attribute using Paths
+		paths, d := resp.State.PathMatches(ctx, path.MatchRoot(resourceName).AtName("actions").AtName(attrName))
+		if d.HasError() {
+			ds.Append(d...)
+			return ds
+		}
+		if len(paths) == 0 {
+			tflog.Info(ctx, "no paths found", map[string]interface{}{
+				"path": fmt.Sprintf("%s.actions.%s", resourceName, attrName),
+			})
+			return ds
+		}
+
+		// set attribute to null using the found Path
+		ds.Append(resp.State.SetAttribute(ctx, paths[0], types.SetNull(usersGroupsResourceModelAttributeTypes))...)
+		if ds.HasError() {
+			return ds
+		}
+	}
+
+	return ds
 }
 
 func (r *permissionResource) UpgradeState(ctx context.Context) map[int64]resource.StateUpgrader {
@@ -318,12 +364,34 @@ func (r *permissionResource) UpgradeState(ctx context.Context) map[int64]resourc
 				}
 
 				resp.Diagnostics.Append(resp.State.Set(ctx, upgradedStateData)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
 
-				setUsersGroupsAttributeToNull(ctx, upgradedStateData.Artifact, "artifact", resp)
-				setUsersGroupsAttributeToNull(ctx, upgradedStateData.Build, "build", resp)
-				setUsersGroupsAttributeToNull(ctx, upgradedStateData.ReleaseBundle, "release_bundle", resp)
-				setUsersGroupsAttributeToNull(ctx, upgradedStateData.Destination, "destination", resp)
-				setUsersGroupsAttributeToNull(ctx, upgradedStateData.PipelineSource, "pipeline_source", resp)
+				resp.Diagnostics.Append(setUsersGroupsAttributeToNull(ctx, upgradedStateData.Artifact, "artifact", resp)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				resp.Diagnostics.Append(setUsersGroupsAttributeToNull(ctx, upgradedStateData.Build, "build", resp)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				resp.Diagnostics.Append(setUsersGroupsAttributeToNull(ctx, upgradedStateData.ReleaseBundle, "release_bundle", resp)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				resp.Diagnostics.Append(setUsersGroupsAttributeToNull(ctx, upgradedStateData.Destination, "destination", resp)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
+
+				resp.Diagnostics.Append(setUsersGroupsAttributeToNull(ctx, upgradedStateData.PipelineSource, "pipeline_source", resp)...)
+				if resp.Diagnostics.HasError() {
+					return
+				}
 			},
 		},
 	}

--- a/pkg/platform/resource_permission_test.go
+++ b/pkg/platform/resource_permission_test.go
@@ -521,6 +521,79 @@ func TestAccPermission_empty_users_state_migration(t *testing.T) {
 	})
 }
 
+func TestAccPermission_no_users_state_migration(t *testing.T) {
+	_, fqrn, permissionName := testutil.MkNames("test-permission", "platform_permission")
+	_, _, groupName := testutil.MkNames("test-group", "artifactory_group")
+
+	temp := `
+	resource "artifactory_group" "{{ .groupName }}" {
+		name = "{{ .groupName }}"
+	}
+
+	resource "platform_permission" "{{ .name }}" {
+		name = "{{ .name }}"
+
+		artifact = {
+			actions = {
+				groups = [
+					{
+						name = artifactory_group.{{ .groupName }}.name
+						permissions = ["READ"]
+					}
+				]
+			}
+
+			targets = []
+		}
+	}`
+
+	testData := map[string]string{
+		"name":      permissionName,
+		"groupName": groupName,
+	}
+
+	config := testutil.ExecuteTemplate(permissionName, temp, testData)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccCheckPermissionDestroy(fqrn),
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"artifactory": {
+						Source: "jfrog/artifactory",
+					},
+					"platform": {
+						Source:            "jfrog/platform",
+						VersionConstraint: "1.7.2",
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "name", testData["name"]),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.users.#", "0"),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.groups.#", "1"),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+			{
+				Config:                   config,
+				ProtoV6ProviderFactories: testAccProviders(),
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"artifactory": {
+						Source: "jfrog/artifactory",
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "name", testData["name"]),
+					resource.TestCheckNoResourceAttr(fqrn, "artifact.actions.users"),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.groups.#", "1"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccPermission_empty_groups_state_migration(t *testing.T) {
 	_, fqrn, permissionName := testutil.MkNames("test-permission", "platform_permission")
 	_, _, userName := testutil.MkNames("test-user", "artifactory_managed_user")
@@ -740,6 +813,81 @@ func TestAccPermission_empty_groups_state_migration(t *testing.T) {
 					resource.TestCheckNoResourceAttr(fqrn, "destination.actions.groups"),
 					resource.TestCheckResourceAttr(fqrn, "pipeline_source.actions.users.#", "1"),
 					resource.TestCheckNoResourceAttr(fqrn, "pipeline_source.actions.groups"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccPermission_no_groups_state_migration(t *testing.T) {
+	_, fqrn, permissionName := testutil.MkNames("test-permission", "platform_permission")
+	_, _, userName := testutil.MkNames("test-user", "artifactory_managed_user")
+
+	temp := `
+	resource "artifactory_managed_user" "{{ .userName }}" {
+		name = "{{ .userName }}"
+		email = "{{ .userName }}@tempurl.org"
+		password = "Password!123"
+	}
+
+	resource "platform_permission" "{{ .name }}" {
+		name = "{{ .name }}"
+
+		artifact = {
+			actions = {
+				users = [
+					{
+						name = artifactory_managed_user.{{ .userName }}.name
+						permissions = ["READ"]
+					}
+				]
+			}
+
+			targets = []
+		}
+	}`
+
+	testData := map[string]string{
+		"name":     permissionName,
+		"userName": userName,
+	}
+
+	config := testutil.ExecuteTemplate(permissionName, temp, testData)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccCheckPermissionDestroy(fqrn),
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"artifactory": {
+						Source: "jfrog/artifactory",
+					},
+					"platform": {
+						Source:            "jfrog/platform",
+						VersionConstraint: "1.7.2",
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "name", testData["name"]),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.users.#", "1"),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.groups.#", "0"),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+			{
+				Config:                   config,
+				ProtoV6ProviderFactories: testAccProviders(),
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"artifactory": {
+						Source: "jfrog/artifactory",
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "name", testData["name"]),
+					resource.TestCheckResourceAttr(fqrn, "artifact.actions.users.#", "1"),
+					resource.TestCheckNoResourceAttr(fqrn, "artifact.actions.groups"),
 				),
 			},
 		},

